### ISSUE TITLE
Update SnakeYaml to 1.31+ due to CVE-2022-25857 / CVE-2022-38750 / CVE-2022-38751

### ISSUE THREAD

**Issue by chadlwilson** [jruby/jruby#7342]
Currently latest JRuby comes bundled with Snakeyaml `1.28` that includes a recently reported DoS vulnerability [CVE-2022-25857](https://nvd.nist.gov/vuln/detail/CVE-2022-25857) / [CVE-2022-38750](https://nvd.nist.gov/vuln/detail/CVE-2022-38750) / https://nvd.nist.gov/vuln/detail/CVE-2022-38751 discovered by OSS Fuzz.

This can be detected by OWASP Dependency Check with checks inside the jars and finds the bundled snakeyaml along with stdlib, e.g: `go/.gradle/caches/modules-2/files-2.1/org.jruby/jruby-complete/9.3.7.0/92b318807fd957ace836078a1df5810e5471f42/jruby-complete-9.3.7.0.jar/META-INF/jruby.home/lib/ruby/stdlib/org/yaml/snakeyaml/1.28/snakeyaml-1.28.jar`

These have been resolved in SnakeYaml `1.31` with a [limit on nested collections](https://bitbucket.org/snakeyaml/snakeyaml/commits/fc300780da21f4bb92c148bc90257201220cf174) and other tweaks.

[Full changelog since 1.28 here](https://bitbucket.org/snakeyaml/snakeyaml/wiki/Changes).

**Expected Behavior**

* Latest JRuby 9.2 and 9.3 come bundled with snakeyaml version which does not report vulnerabilities

**Actual Behavior**

* Latest JRuby `9.3.7.0` and `9.2.21.0` (at time of writing) come bundled with SnakeYaml `1.28` and `1.26` respectively

**Comment by chadlwilson** [comment]
After a bit of digging, looks like this comes as a transitive dependency from https://github.com/ruby/psych and needs updating there.

**Comment by headius** [comment]
@chadlwilson Thanks for the footwork on this. I will assist on the Psych side.

**Comment by chadlwilson** [comment]
@headius How would you like to approach this for JRuby 9.2?

Currently JRuby 9.2 appears to use Psych `3.2.0` and SnakeYAML `1.26` so with ignorance to which "ruby-bits" JRuby relies on from psych, updating to `3.3.3` would be bringing in
- https://github.com/ruby/psych/compare/v3.2.0...v3.3.3
- https://bitbucket.org/snakeyaml/snakeyaml/wiki/Changes

**Comment by chadlwilson** [comment]
OK. Is the ability to upgrade psych without a JRuby release different between `9.2` and `9.3` or does it depend on which mechanism/jars you're using to run JRuby (or whether you are embedding it in a separate Java app).?

**Comment by enebo** [comment]
I also think this is solved by updating pysch without needing a release.  9.3 is only different in that we will be releasing another point release somewhat soon so it will just pick up the new psych at that point.  People using 9.3 today can just update the gem like 9.2 users today.

**Comment by chadlwilson** [comment]
That sounds interesting but I'm not sure I understand how to do that for something like this that is baked into jruby-complete.jar or jruby-stdlib.jar and used as-is as an immutable dependency.

Can you point me in the right direction of docs of how you'd do this?

**Comment by headius** [comment]
Hmmm you make a good point. Neither of those jars, nor the jruby-jars gem can easily be updated.

**Comment by chadlwilson** [comment]
OK - i personally I am happy to close this but will leave it for you to decide whether you want to address for `9.2` or not first.

For others reading, worth noting that there are two other reported "DoS" vulnerabilities currently being reported against all SnakeYaml versions including `1.31`. These are disputed by the authors, who contend they are false positives or at best low priority. Generally these just seem to point to the state state of affairs with respect to how easily it is to get noise into NIST NVD/MITRE as noted [here](https://bitbucket.org/snakeyaml/snakeyaml/wiki/CVE%20&%20NIST). 
- [CVE-2022-38749](https://nvd.nist.gov/vuln/detail/CVE-2022-38749) / https://bitbucket.org/snakeyaml/snakeyaml/issues/525/got-stackoverflowerror-for-many-open
- [CVE-2022-38752](https://nvd.nist.gov/vuln/detail/CVE-2022-38752) / https://bitbucket.org/snakeyaml/snakeyaml/issues/531/stackoverflow-oss-fuzz-47081

I wonder if arguably there are some `LoaderOptions` that should be set in `psych` or able to be set by users to disable certain functionality, but will raise that there.

**Comment by headius** [comment]
@chadlwilson We have discussed 9.2 and will take a "wait and see" approach. If users desperately need the fixed jruby-complete and other jars we will reevaluate, but for now we do not plan to do a 9.2 release in response to this issue.

However, we *will* do a release of 9.3.8.0 to update those jars with patched versions. Should be later this week.

> there are two other reported "DoS" vulnerabilities

I weighed in on one of the new "DoS" issues in SnakeYAML and tend to agree with the author; a stack overflow alone is not much of a DoS, especially if it is being caught and raised properly. It also sounds like it represents a (perhaps problematic) feature in YAML itself: that it requires recursive keys to be possible. However, we may want to take steps to disable this unusual feature, since we'll dodge the DoS concerns that way and only lose obscure and rarely-exercised functionality.

Would you be willing to put together a PR for Psych to disable the related features, assuming they both can be disabled without major impact to functionality? I believe at least the ArrayList stack overflow is part of a feature that can be disabled.

We may or may not delay 9.3.8.0 for the additional configuration...

**Comment by chadlwilson** [comment]
Hi @headius - I'd be willing to help. While it'd be useful to be able to change SnakeYAML's default options, when I looked into SnakeYAML it seemed a recursive keys were disabled by default already, based on a naive look at the option names. 

While I didn't look deeply I inferred that the existing setting doesn't cover the particular array case being highlighted and so configurable LosderOptions wouldn't help us in the short term. I now note it's marked as 'fixed' in 1.32 though. What am I missing?

https://bitbucket.org/snakeyaml/snakeyaml/src/eafb23ec31a0babe591c00e1b50e557a5e3f9a1d/src/main/java/org/yaml/snakeyaml/LoaderOptions.java#lines-22

If that reading is accurate
